lib/client.js

```diff
@@ -114,62 +154,98 @@ module.exports = function (dependencies) {
 
     return new Promise ( resolve => {
       request.on("end", () => {
+        try {
+          if (logger.enabled) {
+            logger(`Request ended with status ${status} and responseData: ${responseData}`);
+          }
+
+          if (status === 200) {
+            resolve({ device });
+          } else if (responseData !== "") {
+            const response = JSON.parse(responseData);
+
+            if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
+              this.config.token.regenerate(tokenGeneration);
+              resolve(this.write(notification, device, retryCount + 1));
+              return;
+            } else if (status === 500 && response.reason === "InternalServerError") {
+              this.closeAndDestroySession();
+              let error = new VError("Error 500, stream ended unexpectedly");
```

Comment:
https://travis-ci.com/github/parse-community/node-apn/jobs/399143362 - travis failure seems spurious, could add a generic helper such as assertErrorIsStreamEndedUnexpectedly to expect either error format

Reply:
I've amended. What do you think?

Reply:
The amended changes look good to me, I don't have any new comments on the latest commits to the PR